feat(apigatewayv2): vpc link and private integrations

ayush987goyal opened the pull request:
VpcLink and HttpProxyPrivateIntegration.

References in AWS docs:
[VpcLink](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-vpclink.html)
[Integration-1](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-private.html)
[Integration-2](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-integration.html#cfn-apigatewayv2-integration-connectiontype)

closes #10119 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 36c005c37484d9d3fed26677717fa4de7076edcf
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1f86a0c4-25c8-459d-a1ec-3be5db05acbf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/vpc-link.ts:
```diff
@@ -0,0 +1,127 @@
+   *
+   * @default - no targets. Use `addTargets` to add targets
+   */
+  readonly targets?: elbv2.INetworkLoadBalancer[];
```
```suggestion
  readonly nlbTargets?: elbv2.INetworkLoadBalancer[];
```

nija-at replied:
Are NLBs the only possible targets for a vpc link?

nija-at replied:
I see you've done this per the spec on v1. Is this true for v2 as well?

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/vpc-link.ts:
```diff
@@ -0,0 +1,127 @@
+
+  private renderTargets() {
+    return this.targets
+      .map(nlb => nlb.vpc?.privateSubnets || [])
```
Curious: why only private subnets? (I'm not well versed in NLB, so excuse me if I'm asking the obvious)

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/vpc-link.ts:
```diff
@@ -0,0 +1,127 @@
+  private renderTargets() {
+    return this.targets
+      .map(nlb => nlb.vpc?.privateSubnets || [])
+      .reduce((acc, subnets) => [...acc, ...subnets], [])
```
```suggestion
      .reduce((acc, subnets) => acc.push(...subnets), [])
```

nija-at reviewed (dismissed):
Thanks for submitting this PR. First round of comments.

ayush987goyal commented:
Hi @nija-at ,

I have modified to previous implementation to a more specific one by also adding the usage with `HttpProxyPrivateIntegration`.

Some open things that I need help on:
1. A better example in the readme.
2. A better way to collect targets in `HttpProxyPrivateIntegration`. Currently I have implemented it to take the ARNs directly.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7b8c452197dcce4bcefdd5548ae463c27ce3dbbd
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=60c00aeb-6d73-4759-9407-1885ee17fb05%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/vpc-link.ts:
```diff
@@ -0,0 +1,109 @@
+   * The name used to label and identify the VPC link.
+   * @default - automatically generated name
+   */
+  readonly vpcLinkName?: string;
```
Add `subnets` as an optional property.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/integrations/http-private.ts:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Properties to initialize a new `HttpProxyPrivateIntegration`.
+ */
+export interface HttpProxyPrivateIntegrationProps {
+  /**
+   * The vpc link to be used for the private integration
+   */
+  readonly vpcLink: IVpcLink;
```
Can we make this optional and create a new `VpcLink` as part of the `bind()` operation and relieve the user from having to create this?

The list of VpcLinks created per vpc could be cached in the `HttpApi` construct. If another integration from the same vpc is introduced, reuse the vpcLink that was already created.

The above logic will be skipped if the optional `vpcLink` property is specified.

ayush987goyal replied:
The above one will be a tricky implementation since the current modelling of integration themselves are not as constructs. So we will have to create the `VpcLink` in the `Route` construct where the `bind` method of integrations are called. 

We might then also have to touch the defaultIntegration and HttpApi for the new Route behaviour and would need to see how we do this for non-private integration. Therefore I have decided to go ahead without this change for now and we could take it as part of further enhancements.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/vpc-link.ts:
```diff
@@ -0,0 +1,109 @@
+   *
+   * @param subnets
+   */
+  public addSubnets(...subnets: ec2.ISubnet[]) {
```
Use [`SubnetSelection`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.SubnetSelection.html) instead of `ISubnet`?

ayush987goyal replied:
The `vpc` itself only has `ISubnet` array as the member variable and I couldn't find an easy way to convert one to other. Hence kept it `ISubnet`

nija-at reviewed (dismissed):
Good job on the first iteration of this PR. Most of this code is conformant to the CDK style and how we model this!

I've not gone through the code in detail but some high level comments around code and class organization that will need to be first addressed.

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/integrations/http-private.ts:
```diff
@@ -0,0 +1,51 @@
+   * The ARN of the target resource.
+   * One of ApplicationLoadBalancerListener, NetworkLoadBalancerListener or CloudMapDiscoveryService
+   */
+  readonly targetArn: string;
```
I implemented the suggestion to an extend by creating individual classes for integration with appropriate resources. Currently they are still in the same package.

nija-at replied:
Feel free to only create the ones that you need and are able to test. Folks after you can add the other integration classes.

ayush987goyal commented:
There is still an open question on a way to verify the integration tests stack (which currently just tests the deployment)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 78e583234c56600c6b445a5e9879aa9b6b19a366
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2f2c197e-4b03-422c-a5af-de3bc4cfbd52%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented:
> The above one will be a tricky implementation since the current modelling of integration themselves are 
> not as constructs. So we will have to create the VpcLink in the Route construct where the bind method of 
> integrations are called.

We do this in several places already. Here's one example - https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-lambda-destinations/lib/lambda.ts#L38-L57

> We might then also have to touch the defaultIntegration and HttpApi for the new Route behaviour and 
> would need to see how we do this for non-private integration. Therefore I have decided to go ahead 
> without this change for now and we could take it as part of further enhancements.

I'm not sure I'm following this. Could you explain a bit better?

It's fine if you want this to be done later but I would like to understand what's preventing this better. 

If it was simple, it could be a nice customer experience.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/README.md:
```diff
@@ -198,3 +199,88 @@ with 3 API mapping resources across different APIs and Stages.
+
+### Private Integration
+
+The `VpcLink` and private integration constructs can be used for integration of `HTTP` APIs with resource behind an Amazon VPC.
+
+#### VPC Link
+
+The following code uses `VpcLink` to connect an HTTP API endpoint to resources in a VPC - 
```
```suggestion
Private integrations let HTTP APIs connect with AWS resources that are placed behind a VPC. These are usually Application
Load Balancers, Network Load Balancers or a Cloud Map service. The `VpcLink` construct enables
this integration.
The following code creates a `VpcLink` to a private VPC.
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/README.md:
````diff
@@ -198,3 +199,88 @@ with 3 API mapping resources across different APIs and Stages.
+Any existing `VpcLink` resource can be imported into the CDK app via the `VpcLink.fromVpcLinkId()`.
+
+```ts
+const stack = new Stack(app, 'my-stack');
+
````
```suggestion
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/README.md:
````diff
@@ -198,3 +199,88 @@ with 3 API mapping resources across different APIs and Stages.
+});
+```
+
+#### ServiceDiscoveryPrivateIntegration
````
```suggestion
#### Service Discovery
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/integration.ts:
```diff
@@ -165,6 +195,20 @@ export interface HttpRouteIntegrationConfig {
   readonly method?: HttpMethod;
 
+  /**
+   * The ID of the VPC link for a private integration. Supported only for HTTP APIs.
```
```suggestion
   * The ID of the VPC link for a private integration.
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/integration.ts:
```diff
@@ -165,6 +195,20 @@ export interface HttpRouteIntegrationConfig {
+  /**
+   * The ID of the VPC link for a private integration. Supported only for HTTP APIs.
+   *
+   * @default - undefined
```
```suggestion
   * @default - no vpc link integration
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/integrations/private-integration/application-loadbalancer-integration.ts:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Properties to initialize `ApplicationLoadBalancerPrivateIntegration`.
+ */
+export interface ApplicationLoadBalancerPrivateIntegrationProps extends BasePrivateIntegrationProps {
```
As mentioned here - https://github.com/aws/aws-cdk/pull/10531#discussion_r500864507 - these will need to be spun off into a new module `aws-apigatewayv2-integrations`.

To keep the scope contained, implement only private integration the one that you need, but it's best to set up the correct pattern for those to follow.

Let me know how I can help with this.

ayush987goyal replied:
Done!

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/integrations/private-integration/http-private.ts:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Common base props for private integration
+ */
+export interface BasePrivateIntegrationProps {
```
Avoid exporting interfaces that are internal

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/integrations/private-integration/application-loadbalancer-integration.ts:
```diff
@@ -0,0 +1,25 @@
+/**
+ * The Application Load Balancer Private integration resource for HTTP API
+ */
+export class ApplicationLoadBalancerPrivateIntegration extends HttpProxyPrivateIntegration {
```
Some terser names would be nicer.

```suggestion
export class ALBIntegration extends HttpProxyPrivateIntegration {
```

ayush987goyal replied:
Done!

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/README.md:
````diff
@@ -198,3 +199,88 @@ with 3 API mapping resources across different APIs and Stages.
+const awesomeLink = VpcLink.fromVpcLinkId(stack, 'awesome-vpc-link', 'us-east-1_oiuR12Abd');
+```
+
+#### ApplicationLoadBalancerPrivateIntegration
+
+The following code is a basic application load balancer private integration of HTTP API:
````
This and the following sections should go into the readme for the new module `aws-apigatewayv2-integrations`.

ayush987goyal replied:
Done!

nija-at reviewed (dismissed):
See my comments below. The main one is about splitting off into its own package.

Also, take a look at this one - https://github.com/aws/aws-cdk/pull/10531#issuecomment-707000111

nija-at commented:
> There is still an open question on a way to verify the integration tests stack (which currently just tests the deployment)

You can run the integ test using `yarn integ integ.custom-domain.js`. This will deploy and then destroy the stack. If all goes well, it will report success.

Credentials can be configured using any means that you do for the AWS CLI.

ayush987goyal commented:
> > There is still an open question on a way to verify the integration tests stack (which currently just tests the deployment)

> You can run the integ test using yarn integ integ.custom-domain.js. This will deploy and then destroy the stack. If all goes well, it will report success.

> Credentials can be configured using any means that you do for the AWS CLI.

Yeah I was able to test the deployment successfully. I was specifically talking about the verification steps we mention at the top of the integ-tests file. I was looking for a way to verify these stacks basically.

nija-at commented:
> Yeah I was able to test the deployment successfully. I was specifically talking about the verification steps we mention at the top of the integ-tests file. I was looking for a way to verify these stacks basically.

We currently don't have a way to run those tests in the CDK. These should contain steps on verification beyond deployment. 

As an example, in your case, it should deploy a stack that creates an ALB that returns '200' status code, which is then integrated to an `HttpApi` construct. The steps should simply be call `curl` on the API endpoint and it should return '200'. 
This internally will verify that the request goes to the ALB successfully and the response is correctly received. 

Hope that makes it clear.

ayush987goyal commented:
Hi @nija-at ,

I have implemented the integrations by taking into account your suggestions and separated out the packages as well.

The `ServiceDiscoveryPrivateIntegration` still requires a `vpcLink` to be supplied mandatorily (unlike the ALB and NLB integrations) since there is no way to extract out the VPC from the `Serivce` or the `Namespace`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: faa7c50a0df1c02d2ff5aaac5119e8dd2f9654b6
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f1f0194a-3021-4a52-ac62-d52e6f8b8006%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9e990b7afeaaf8aa5756247168bbb4fc2df54e27
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ba838ddc-a219-40ce-b96a-fae97628b424%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b0234344f90b208110acebb3b022f1a215de3140
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5ed68dbf-92dd-4cf5-9150-6c02e53f6080%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 168fdb98585d996bb2c82da43bc3d2a315ff0fcf
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6b211dc6-3e6e-47dd-b45e-4a78aa48f073%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: fd51daed1151e25f8793b8c99747dfb1405ae0dd
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9e5687e2-5679-4167-957b-94ea7b9ac7f9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/README.md:
```diff
@@ -0,0 +1,83 @@
+
+## Introduction
+
+API Gateway can be integrated with several other AWS services. Curently the following integrations are supported
+by this integrations construct library.
```
```suggestion
Integrations connect a route to backend resources. HTTP APIs support Lambda proxy, AWS service, and HTTP proxy integrations. HTTP proxy integrations are also known as private integrations.

Currently the following integrations are supported by this construct library.
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/README.md:
```diff
@@ -0,0 +1,83 @@
+
+## Private Integration
+
+Following private integrations are supported by API Gateway for resources behind an Amazon VPC.
```
```suggestion
Private integrations enable integrating an HTTP API route with private resources in a VPC, such as Application Load Balancers or
Amazon ECS container-based applications.  Using private integrations, resources in a VPC can be exposed for access by
clients outside of the VPC.

The following integrations are supported for private resources in a VPC.
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/README.md:
```diff
@@ -0,0 +1,83 @@
+
+Following private integrations are supported by API Gateway for resources behind an Amazon VPC.
+
+#### ALBPrivateIntegration
```
```suggestion
### Application Load Balancer
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/README.md:
````diff
@@ -0,0 +1,83 @@
+});
+```
+
+#### NLBPrivateIntegration
````
```suggestion
### Network Load Balancer
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/README.md:
````diff
@@ -0,0 +1,83 @@
+});
+```
+
+#### ServiceDiscoveryPrivateIntegration
````
```suggestion
### Cloud Map Service Discovery
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/README.md:
```diff
@@ -24,6 +24,7 @@
   - [Publishing HTTP APIs](#publishing-http-apis)
   - [Custom Domain](#custom-domain)
   - [Metrics](#metrics)
+  - [VPC Link](#vpc-link)
```
Add a section after vpc link called 'private integration', mention an introduction sentence on it and then point folks at the new module.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/lib/alb-integration.ts:
```diff
@@ -0,0 +1,40 @@
+import { HttpMethod, IVpcLink } from '@aws-cdk/aws-apigatewayv2';
+import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
+import { HttpProxyPrivateIntegration } from './http-private';
+
```
apigatewayv2 supports both HTTP API and Websocket APIs. As far as I can tell, all of this new integration is aimed at HTTP APIs only and will not work for websocket.

Given this, move all of these under `lib/http/` folder and perhaps add `Http` as a prefix so that these classes are namespaced and won't collide with potential future integrations for Websocket.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/lib/alb-integration.ts:
```diff
@@ -0,0 +1,40 @@
+/**
+ * The Application Load Balancer Private integration resource for HTTP API
+ */
+export class ALBPrivateIntegration extends HttpProxyPrivateIntegration {
```
Naming: 
1. We use pascal case for acronyms, so this should be 'Alb' and not 'ALB'.

2. Terser - Just `AlbIntegration` should be sufficient. The "private" feels redundant. Incorporating the previous comment, this will be `HttpAlbIntegration`.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/lib/http-private.ts:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Properties to initialize a new `HttpProxyPrivateIntegration`.
+ */
+export interface HttpProxyPrivateIntegrationProps {
```
Same about naming here: can this just be `HttpPrivateIntegrationProps` and `HttpPrivateIntegration`?

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/lib/http-private.ts:
```diff
@@ -0,0 +1,63 @@
+  readonly vpcLink?: IVpcLink;
+
+  /**
+   * The vpc for which the VpcLink needs to be created
+   *
+   * @default - undefined
```
What does it mean for this to be left undefined? When would a customer not specify this?

ayush987goyal replied:
The `vpc` and `vpcLink` props are both optional with at least one of them needs to be present in the props (we create a `vpcLink` if it is not provided but vpc is).

nija-at replied:
Ok, makes sense.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/lib/http-private.ts:
```diff
@@ -0,0 +1,63 @@
+      method: this.props.method ?? HttpMethod.ANY,
+      payloadFormatVersion: PayloadFormatVersion.VERSION_1_0, // 1.0 is required and is the only supported format
+      type: HttpIntegrationType.HTTP_PROXY,
+      connectionId: this.props.vpcLink?.vpcLinkId,
```
Don't you have to create a new vpc link here, if `undefined`?

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/lib/alb-integration.ts:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Properties to initialize `ALBPrivateIntegration`.
+ */
+export interface ALBPrivateIntegrationProps {
+  /**
+   * The listener to the application load balancer used for the integration
+   */
+  readonly listener: elbv2.ApplicationListener;
+
+  /**
+   * The vpc link to be used for the private integration
+   *
+   * @default - a new VpcLink is created
+   */
+  readonly vpcLink?: IVpcLink;
+
+  /**
+   * The HTTP method that must be used to invoke the underlying HTTP proxy.
+   * @default HttpMethod.ANY
+   */
+  readonly method?: HttpMethod;
+}
```
I think you can just extend `HttpProxyPrivateIntegrationProps` and drop everything except for the `listener` property.

ayush987goyal replied:
We cannot directly extend the `HttpPrivateIntegrationProps` interface because of required `targetArn` prop. Created a `BasePrivateIntegrationProps` interface for the same.

Note: The `HttpServiceDiscoveryIntegration` integration does not extend this because of the required `vpcLink` prop.

nija-at replied:
I'm not a fan of the `BasePrivateIntegrationProps`, especially because it needs to exported publicly. No value in exposing this to customers and increasing the surface area.

I think I know what's going on here. The root problem is that the integration between `HttpPrivateIntegration` and any of its subclasses is through the constructor.

How about instead changing this to be via `bind()`? Technically, each implementation of `IHttpRouteIntegration` should re-implement `bind()` suitable to it. This would look something like - 

```ts
export abstract class HttpPrivateIntegration implements IHttpRouteIntegration {
  protected configureVpcLink(bindOptions: HttpRouteIntegrationBindOptions, integrationOptions: HttpPrivateIntegrationOptions): VpcLink {
    // ...
    return options.route.httpApi.addVpcLink(...);
  }

  public abstract bind(options: HttpRouteIntegrationBindOptions);
}

export class HttpAlbIntegration extends HttpPrivateIntegration {
  constructor(private readonly props: HttpAlbIntegrationProps) {}

  public abstract bind(options: HttpRouteIntegrationBindOptions) {
    const vpclink = this.configureVpcLink(options, props);
    return {
      ...
    }
  }
}
```

With this form of structuring, it should be possible for `HttpAlbIntegrationProps` to extend `HttpPrivateIntegrationOptions` (formerly: `HttpPrivateIntegrationProps`).

An added benefit is that, each specific implementation has a lot more control over how it's bound. Any commonality like `PayloadFormatVersion` can be captured at the base class but can be overridden by a specific implementation if it wishes to.

Does this make sense?

nija-at replied:
I'll go one step further and say, we shouldn't even export `HttpPrivateIntegration`.

A [limitation of JSII](https://github.com/aws/jsii/issues/2159) would mean that all sub-classes will need to be in the same file, but looking at the size of these classes, this might be ok.

nija-at replied:
Updated my comment above

ayush987goyal replied:
Yes that makes sense and is better way to implement. Will make the appropriate changes.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/route.ts:
```diff
@@ -120,16 +120,28 @@ export class HttpRoute extends Resource implements IHttpRoute {
+        throw new Error('One of connectionId or vpc should be provided for private integration');
+      }
+
+      const vpcLink = this.httpApi.addVpcLink({ vpc: config.vpc });
```
Can we call `addVpcLink()` from the `bind()` method of the base class `HttpProxyPrivateIntegration` instead? Feels like the right layer of responsibility and return the resultant vpcLink as the result of the bind.

You should be able to go from `HttpRouteIntegrationBindOptions` → `IHttpRoute` → `IHttpApi`.

nija-at reviewed (dismissed):
Wow! IMPRESSIVE.

Thanks for all the work here. I'm sure you had a few hoops to jump through.

Mostly adjustments to README and naming conventions, besides one or two other comments. This has been great!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2efea15434f55e5520efca74b6f9173081c1404e
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3d0fd7ad-5ca5-4965-960a-bfd2298ae071%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 036824d7e99c8c82d104b0b10258614c8d1065c5
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=15db80f1-0328-4d38-9f42-3f8129212062%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: bc759877ac190e3c9fa32310dccb13ce0427d732
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=31330846-750b-46d0-8e8d-6f291df7cc19%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 900c701718f258978b6beb4046040f36cc9fd954
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b5695f4c-827e-4438-8c3f-29b44d6d700a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/lib/http/http-private.ts:
```diff
@@ -0,0 +1,76 @@
+  public bind(options: HttpRouteIntegrationBindOptions): HttpRouteIntegrationConfig {
+    let vpcLink = this.props.vpcLink;
+    if (!vpcLink) {
+      if (!this.props.vpc) {
+        throw new Error('One of vpcLink or vpc should be provided for private integration');
+      }
```
error when both are provided and they don't point to the same vpc

ayush987goyal replied:
Is that validation required? We are currently ignoring the passed vpc if the vpcLink is supplied. Basically vpcLink trumps the vpc.

Also, the vpcLink in v2 can point to multiple VPCs in theory (since it takes subnets and securityGroups). In this case it might not make sense to compare the VPCs.

Thoughts?

nija-at replied:
Accepting a property and not using it is usually confusing. "Why was my vpc not used?" would be a complaint I expect to hear.

How about just failing if both are presented? It'll make the requirement that only one should be passed clear.

ayush987goyal replied:
Yeah, sounds good.

ayush987goyal replied:
This should not be required anymore since we are not accepting vpc as a prop (it is passed around internally by integrations)

nija-at replied:
How come? 
In the case of `HttpAlbIntegration`, we accept an ALB and a VPCLink. If the ALB is in a different vpc from the vpc link, it's useful to error, no?

ayush987goyal replied:
Oh that way. 

But in that case the intent was explicit to use the passed vpcLink right? We are not explicitly taking the vpc as a prop. We are inferring.

Also, as mentioned before, the vpcLink in v2 is not bound to just one vpc in theory. It revolves around subnets and securityGroups. We have just provided a convenience to pass a vpc to create the vpcLink. So comparing the vpc of vpcLink and ALB listener might not be ideal.

Thoughts?

nija-at replied:
I see. Ok, let's leave it as is for now. See how this pans out.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/lib/http/service-discovery-integration.ts:
```diff
@@ -0,0 +1,37 @@
+   */
+  readonly service: servicediscovery.Service;
+
+  /**
+   * The vpc link to be used for the private integration
+   */
+  readonly vpcLink: IVpcLink;
```
curious why `vpcLink` is mandatory here and cannot use the same logic as the other two?

ayush987goyal replied:
There is currently no way to extract out the underlying vpc from the `Service` construct. It should ideally come from
Service -> Namespace -> VPC
This is currently not present in the discovery module. We can update the current implementation when we have that.

nija-at replied:
That's fine with me.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9e1b5663c62a165658f42d0f7be947dc6edb1cb4
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=efece2e8-5707-497b-8527-554333573d2f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented:
I've taken the liberty to pare down the file names to make it more terse and moved the `http-private.ts` into a `private/` folder, like we do in other parts of the CDK construct library. Let me know if you're ok with this.

The code looks great otherwise.

ayush987goyal commented:
Thanks for these changes. They look good to me.

nija-at reviewed (approved):
All the good stuff 🙌

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 99432048c241d62620f599aaf429c01d16d55b08
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=62e2f06b-f1f2-4a73-bd6f-f930c81a85ba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3855ada8948beb9a4f50e7bb266f5798a6629de7
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8cacdabb-cceb-498a-b95b-32996a987181%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented:
heads up: I believe this also requires aws/jsii#2172 as the usage of `@internal` is causing downstream compilation failures currently.

may need to temporarily revert this.

ayush987goyal commented:
@shivlaks @nija-at Please go ahead with the solution you seem fit.

shivlaks commented:
> @shivlaks @nija-at Please go ahead with the solution you seem fit.

@ayush987goyal perfect. we're going to temporarily revert this and merge this back in when the next release of `jsii` is consumed in the cdk. stay tuned!
